packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts

```diff
@@ -183,6 +232,90 @@ export default class UserStorageController extends BaseController<
     },
   };
 
+  #accounts = {
+    // This is replaced with the actual value in the constructor
+    // We will remove this once the feature will be released
+    isAccountSyncingEnabled: false,
+    setIsUserStorageAccountSyncingInProgress: async (
+      isUserStorageAccountSyncingInProgress: boolean,
+    ) => {
+      // Publish event
+      const eventToPublish = isUserStorageAccountSyncingInProgress
+        ? 'UserStorageController:accountSyncingInProgress'
+        : 'UserStorageController:accountSyncingComplete';
+
+      this.messagingSystem.publish(
+        eventToPublish,
+        isUserStorageAccountSyncingInProgress,
+      );
+
+      // Update state
+      this.update((state) => {
+        state.isUserStorageAccountSyncingInProgress =
+          isUserStorageAccountSyncingInProgress;
+      });
+    },
+    getInternalAccountByAddress: async (address: string) => {
+      return this.messagingSystem.call(
+        'AccountsController:getAccountByAddress',
+        address,
+      );
+    },
+    getInternalAccountsList: async (): Promise<InternalAccount[]> => {
+      return this.messagingSystem.call('AccountsController:listAccounts');
+    },
+    getUserStorageAccountsList: async (): Promise<
+      UserStorageAccount[] | null
+    > => {
+      const rawAccountsListResponse =
+        await this.performGetStorageAllFeatureEntries('accounts');
+
+      return (
+        rawAccountsListResponse?.map((rawAccount) => JSON.parse(rawAccount)) ??
+        null
+      );
+    },
+    saveInternalAccountToUserStorage: async (address: string) => {
+      const internalAccount = await this.#accounts.getInternalAccountByAddress(
+        address,
+      );
+
+      if (!internalAccount) {
+        return;
+      }
+
+      // Map the internal account to the user storage account schema
+      const mappedAccount =
+        mapInternalAccountToUserStorageAccount(internalAccount);
+
+      await this.performSetStorage(
+        `accounts.${address}`,
+        JSON.stringify(mappedAccount),
+      );
+    },
+    saveInternalAccountsListToUserStorage: async () => {
+      const internalAccountsList =
+        await this.#accounts.getInternalAccountsList();
+
+      if (!internalAccountsList) {
+        return;
+      }
+
+      const userStorageAccountsList = internalAccountsList.map(
+        mapInternalAccountToUserStorageAccount,
+      );
+
+      await Promise.all(
+        userStorageAccountsList.map(async (userStorageAccount) => {
+          await this.performSetStorage(
+            `accounts.${userStorageAccount.a}`,
+            JSON.stringify(userStorageAccount),
+          );
+        }),
```

Comment:
Future - asked the BE team to provide a batch endpoint so we don't need to perform multiple API calls here.